ts/Series/Treegraph/TreegraphLayout.ts

```diff
@@ -0,0 +1,419 @@
+        });
+        node.relativeXPosition = index;
+    },
+    /**
+     * Recursive post order traversal of the tree, where the initial position
+     * of the nodes is calculated.
+     *
+     * @param {TreegraphNode} node the node for which the position should be
+     * calculated
+     */
+    firstWalk: function (
+        this: Highcharts.TreegraphLayout,
+        node: TreegraphNode
+    ): void {
+        const treeLayout = this;
+        let leftSibling;
+        // Arbitrary value used to position nodes in respect to each other.
+        let siblingDistance = 1;
+        // If the node is a leaf, set it's position based on the left siblings.
+        if (node.isLeaf()) {
+            leftSibling = node.getLeftSibling();
+            if (leftSibling) {
+                node.preX = leftSibling.preX + siblingDistance;
+                node.mod = node.preX;
+            } else {
+                node.preX = 0;
+            }
+        } else {
+            // If the node has children, perform the recursive first walk for
+            // its children, and then calculate its shift in the apportion
+            // function (most crucial part part of the algorythm).
+            let defaultAncestor = node.getLeftMostChild() as TreegraphNode;
+
+            node.linksFrom.forEach(function (link): void {
+                treeLayout.firstWalk(link.toNode);
+                defaultAncestor = treeLayout.apportion(
+                    link.toNode,
+                    defaultAncestor
+                );
+            });
+            treeLayout.executeShifts(node);
+
+            let leftChild = node.getLeftMostChild() as TreegraphNode;
+            let rightChild = node.getRightMostChild() as TreegraphNode;
+            // Set the position of the parent as a middle point of its children
+            // and move it by the value of the leftSibling (if it exists).
+            let midPoint = (leftChild.preX + rightChild.preX) / 2;
+            leftSibling = node.getLeftSibling();
+
+            if (leftSibling) {
+                node.preX = leftSibling.preX + siblingDistance;
+                node.mod = node.preX - midPoint;
+            } else {
+                node.preX = midPoint;
+            }
+        }
+    },
+    /**
+     * Pre order traversal of the tree, which sets the final xPosition of the
+     * node as its preX value and sum of all if it's parents' modifiers.
+     *
+     * @param {TreegraphNode} node the node, for which the final position
+     * should be calculated.
+     * @param {number} modSum The sum of modifiers of all of the parents.
+     */
+    secondWalk: function (
+        this: Highcharts.TreegraphLayout,
+        node: TreegraphNode,
+        modSum: number
+    ): void {
+        const treeLayout = this;
+        // When the chart is not inverted we want the tree to be positioned from
+        // left to right with root node close to the chart border, this is why
+        // x and y positions are switched.
```

Comment:
Q: Shouldn't we somehow check if the chart is inverted then?

Reply:
By naming the variables (e.g. leftIntNode, preX, getLeftChild, etc) the algorithm assumes, that the layout will be positioned from top to bottom, and the default (non-iverted) layout of the nodes in Sankey and organization are from left to right. 
I decided to continue the trend, and the easiest way of switching it was on the last step so that the algorithm assigns values that are consistent with our orientation.

Reply:
Ah, okay, thanks for the explanation 👍